game/src/js/scenes/GamepadScene.js

```diff
@@ -0,0 +1,91 @@
+import BaseScene from '../scenes/BaseScene';
+
+export default class GamepadScene extends BaseScene {
+  constructor() {
+    super({ key: 'GamepadScene', active: true });
+  }
+
+  preload() {
+    this.load.image('joystick', '/assets/sprites/joystick.png');
+    this.load.image('joystick_base', '/assets/sprites/joystick_base.png');
+  }
+
+  create() {
+    const { height } = this.cameras.main;
+    this.MIN_DIST = 5;
+    this.MAX_DIST = 40;
+    this.START_X = 100;
+    this.START_Y = height / 2;
+
+    this.joystickBase = this.add
+      .image(this.START_X, this.START_Y, 'joystick_base')
+      .setAlpha(0.5);
+    this.joystick = this.add
+      .image(this.START_X, this.START_Y, 'joystick')
+      .setAlpha(0.5);
+    this.joystickBase.visible = false;
+    this.joystick.visible = false;
+    this.joystick.setInteractive();
+    this.input.setDraggable(this.joystick);
+
+    this.joystick.on('drag', (pointer, x, y) => {
+      const xDist = Math.abs(this.START_X - x);
+      const yDist = Math.abs(this.START_Y - y);
+      this.joystick.x =
+        xDist < this.MAX_DIST
+          ? x
+          : x > this.START_X ? this.START_X + this.MAX_DIST : this.START_X - this.MAX_DIST;
+      this.joystick.y =
+        yDist < this.MAX_DIST
+          ? y
+          : y > this.START_Y ? this.START_Y + this.MAX_DIST : this.START_Y - this.MAX_DIST;
+
+      this.updateController(x, y);
+    });
+
+    this.joystick.on('dragend', () => {
+      this.joystick.x = this.START_X;
+      this.joystick.y = this.START_Y;
+      this.updateController(this.START_X, this.START_Y);
+    });
+  }
+
+  updateController(x, y) {
+    const controller = this.scene.manager.keys.GameScene.controller
+      || this.scene.manager.keys.DungeonScene.controller;
+    const keys = {
+      up: { isDown: false },
+      down: { isDown: false },
+      left: { isDown: false },
+      right: { isDown: false }
+    };
+
+    if (x > this.START_X + this.MIN_DIST) {
+      keys.right.isDown = true;
+    }
+
+    if (x < this.START_X - this.MIN_DIST) {
+      keys.left.isDown = true;
+    }
+
+    if (y > this.START_Y + this.MIN_DIST) {
+      keys.down.isDown = true;
+    }
+
+    if (y < this.START_Y - this.MIN_DIST) {
+      keys.up.isDown = true;
+    }
+
+    if (controller) {
+      controller.setVirtualKeys(keys);
+    }
+  }
+
+  setVisible(visible) {
+    const { height, width } = this.cameras.main;
+    if (Math.max(height, width) < 768) {
```

Comment:
Maybe a better check here would be to determine if the user is on a mobile device. Something like: `if(/Mobi/i.test(navigator.userAgent) || /Android/i.test(navigator.userAgent))` should do it. Otherwise, large tablets may not show these controls.

Reply:
Yeah that's a great idea